### PR TITLE
Column Statistics

### PR DESCRIPTION
**Pull request by JCOR11599**
Gathers the following column stats:
 - mean
 - min
 - max
 - count
 - null count
 - distinct count

and stores them in a modified `information_schema.column_statistics` table.

There is now a new `sql.Histogram` struct that contains the column stats; the contents and format of `sql.Histogram` come from a mixture of Cockroach and MySQL.

Additionally, there is also a TableStatistics interface, which is meant to be implemented if people want statistics on that table.

### CODE REVIEW

File: sql/mysql_db/column_statistics.go
Comment (max-hoffman) [inline sql/mysql_db/column_statistics.go:100]: there aren't any ref params, should be fine

File: sql/mysql_db/column_statistics_table.go
Comment (max-hoffman) [inline sql/mysql_db/column_statistics_table.go:32]: these should probably be `errors.New` and in the `errors.go` file

File: sql/plan/analyze.go
Comment (max-hoffman) [inline sql/plan/analyze.go:81]: copy(dest, src) might be more concise/faster

File: sql/plan/analyze.go
Comment (max-hoffman) [inline sql/plan/analyze.go:112]: these might be helpful while dev'ing, but if I am being picky i'm not sure all of these comments are value-additive afterwards. if you want to add docs describing what methods do at a high level, add near the signature

File: sql/plan/analyze.go
Comment (max-hoffman) [inline sql/plan/analyze.go:117]: i'm not sure if they can reach here, but `IndexedTableAccess`, `DeferredAsofTable`, and maybe a couple others are valid table types. A prepared enginetest would probably break, for example.

**Review by max-hoffman — APPROVED** [review]
This is directionally good, OK first pass. Needs a lot of enginetests to verify all of the schemas/read/write/mutate behavior you expect to support here before merging. The statistics flatbuffer maybe needs to be replaced with an interface that lets integrators store stats in a table.

File: sql/core.go
Comment (max-hoffman) [inline sql/core.go:426]: is there a reason these need to be interface? seems like they could just be a struct

File: sql/errors.go
Comment (max-hoffman) [inline sql/errors.go:578]: more specific error names pls

**Review by max-hoffman — COMMENTED** [review]
That was a quick turnaround! Next steps and then I think this is ready for merge:
- statistic intermediary can be a struct, its preferable if every integrator uses one `Histogram` rather than implementing their own
- I suspect we only need histogram, and that the other fields are redundant. A histogram is a struct with buckets (also struct). Each bucket has a range bounds, null count, distinct count (null and distinct may be premature, but PG and CR uses). For example, a `HistogramTable` might have `HistogramForIndex`, or a `HistogramIndex` would have `Histogram()` depending on which you pick. I think a separate interface, `Coster`, will use histograms to do that actual estimating.
- GMS Analyze shouldn't literally populate a table. It is the right idea, but that logic should go in the integrator. Specifically, remove flatbuffers, add a map to the memory table (or to indexes) `map[string]sql.IndexStatistics` for saving/returning stats when asked. We can start with an in-memory map on the Dolt side also (or omit before closing this PR), and evolve it into a persisting form later. The analyze function body will be small, because the integrator does the work of populating/saving.
- The statistics info schema table should get histogram info the same way the analyzer does, through the tables/catalog

File: sql/plan/analyze.go
Comment (max-hoffman) [inline sql/plan/analyze.go:31]: can other table nodes end up here? do we error earlier if a `QueryProcess` or `DeferredAsofTable` node is here for some reason?

File: sql/plan/analyze.go
Comment (max-hoffman) [inline sql/plan/analyze.go:75]: maybe have a TODO for cross or multi database analyze

File: memory/table.go
Comment (max-hoffman) [reply memory/table.go:361]: GMS should probably expose a Histogram builder? one implementation of a class/function that inputs rows and incrementally builds a balanced histogram? I can't think of whether we would want to do this differently in Dolt vs GMS.

Comment (max-hoffman) [reply memory/table.go:361]: simple form is fine for first pass

**Review by max-hoffman — COMMENTED** [review]
Mostly LGTM but I'd like one last look after you cleanup comments, extraneous fields, some of the interface methods, and other nits. I don't know what's going on with the info schema table, i'll just assume you are abiding by the way other info schema tables work, though extra tests to make sure it doesn't fall apart with schema changes would be nice. Focus comments and TODOs on areas we plan on improving in the near-term

File: memory/table.go
Comment (JCOR11599) [reply memory/table.go:281]: O(2n) = O(n) :)

This was just left over code from the `NumRows` function, will change to not do that.

File: sql/core.go
Comment (JCOR11599) [reply sql/core.go:466]: It is always true after the first successful analyze. I added this to more easily check if the statistics variable in the structs was filled with valid data. 

Additionally, the dolt implementation of this method looks is used to look into the dbState for stats as they aren't cached there.

File: memory/table.go
Comment (JCOR11599) [reply memory/table.go:351]: I think the frequency map I create here is a useful intermediate step in creating histograms for various number of buckets and various ranges.

File: sql/core.go
Comment (zachmu) [inline sql/core.go:443]: Call this AnalyzeTable

Comment: `AnalyzeTable is a hook to update any cached or persisted statistics for this table. It may be triggered by an analyze table statement, or automatically when the engine decides it is necessary. Integrators can ignore this hook and implement their own method of keeping statistics up to date, at the cost of potentially stale statistics.`

File: sql/core.go
Comment (zachmu) [inline sql/core.go:444]: GetStatistics returns statistics for this table

File: sql/core.go
Comment (zachmu) [inline sql/core.go:424]: TableStatistics provides access to statistical information about the values stored in a table

File: sql/core.go
Comment (zachmu) [inline sql/core.go:426]: Method docs for all these interface methods

File: sql/core.go
Comment (zachmu) [inline sql/core.go:427]: This is redundant with schema and should be removed

File: sql/core.go
Comment (zachmu) [inline sql/core.go:428]: This is redundant with StatisticsTable.NumRows

Recommend getting rid of the latter and just using this

File: sql/errors.go
Comment (zachmu) [inline sql/errors.go:593]: This is a bad error message. What is it trying to convey? Why doesn't it echo the value that was given?

File: sql/errors.go
Comment (zachmu) [inline sql/errors.go:596]: Same comment

File: sql/plan/analyze.go
Comment (zachmu) [inline sql/plan/analyze.go:99]: This panics if the table isn't a sql.StatisticsTable. Catch this and return an error in that case.

File: sql/plan/analyze.go
Comment (zachmu) [inline sql/plan/analyze.go:102]: This is an anti-pattern we're trying to get rid of: rather than doing work in RowIter(), implement a simple iter object and do work in Next()

File: sql/errors.go
Comment (zachmu) [reply sql/errors.go:596]: Both these errors appear to be unused, just get rid of them

**Review by zachmu — COMMENTED** [review]
Mostly looks good, some comments on the interface design

File: sql/core.go
Comment (zachmu) [reply sql/core.go:444]: Just call this `Statistics()`, that's the pattern we use for most of these accessor interface (no `Get`)

File: sql/core.go
Comment (JCOR11599) [reply sql/core.go:428]: There are other things things that rely on `StatisticsTable.NumRows()` like `ShowTableStatus.RowIter()`.
If those now rely on RowCount(), the table needs to be analyzed right before to provide an accurate count.

What's the best way to handle this? Make a new interface with `NumRows()`? Leave it for now?

Comment (zachmu) [reply sql/core.go:428]: This isn't true. An integrator can decide to always return an accurate row count independent of when analyze is invoked.

Keep all the table stats in the same interface, no redundancies.

File: sql/analyzer/inserts.go
Comment (zachmu) [inline sql/analyzer/inserts.go:66]: Got too aggressive with a rename looks like

File: sql/core.go
Comment (zachmu) [inline sql/core.go:431]: Just delete now

File: sql/core.go
Comment (zachmu) [inline sql/core.go:432]: Why case-sensitive? Nothing else is

Comment (zachmu) [reply sql/core.go:432]: Also mention what a nil histogram means

**Review by zachmu — APPROVED** [review]
LGTM

**Pull request merged**
